lib/mlir/Dialect/Clift/IR/ImportModel.cpp

```diff
@@ -0,0 +1,599 @@
+      return getTypeAttribute(T);
+    }
+    case model::TypeKind::RawFunctionType: {
+      RequireComplete = true;
+      auto const &T = llvm::cast<model::RawFunctionType>(ModelType);
+      return getTypeAttribute(T);
+    }
+    case model::TypeKind::StructType: {
+      auto const &T = llvm::cast<model::StructType>(ModelType);
+      return getTypeAttribute(T, RequireComplete);
+    }
+    case model::TypeKind::TypedefType: {
+      auto const &T = llvm::cast<model::TypedefType>(ModelType);
+      return getTypeAttribute(T, RequireComplete);
+    }
+    case model::TypeKind::UnionType: {
+      auto const &T = llvm::cast<model::UnionType>(ModelType);
+      return getTypeAttribute(T, RequireComplete);
+    }
+    default:
+      revng_abort();
+    }
+  }
+
+  RecursiveCoroutine<clift::ValueType>
+  getUnwrappedType(const model::Type &ModelType,
+                   bool RequireComplete = false,
+                   const bool Const = false) {
+    if (const auto *const T = llvm::dyn_cast<model::PrimitiveType>(&ModelType))
+      rc_return getPrimitiveType(*T, Const);
+
+    auto const getDefinedType = [&](const auto Attr) -> clift::ValueType {
+      return make<clift::DefinedType>(Attr, getBool(Const));
+    };
+
+    if (const auto It = Cache.find(ModelType.ID()); It != Cache.end())
+      rc_return getDefinedType(It->second);
+
+    const clift::TypeDefinition Attr = getTypeAttribute(ModelType,
+                                                        RequireComplete);
+
+    if (not Attr)
+      rc_return nullptr;
+
+    if (RequireComplete) {
+      const auto R = Cache.try_emplace(ModelType.ID(), Attr);
+      revng_assert(R.second);
+    }
+
+    rc_return getDefinedType(Attr);
+  }
+
+  RecursiveCoroutine<clift::ValueType>
+  getQualifiedType(const model::QualifiedType &ModelType,
+                   bool RequireComplete = false) {
+    if (not ModelType.UnqualifiedType().isValid()) {
+      if (EmitError)
+        EmitError() << "Invalid UnqualifiedType in QualifiedType";
+      rc_return nullptr;
+    }
+
+    auto Qualifiers = llvm::ArrayRef(ModelType.Qualifiers());
+
+    const auto takeQualifier = [&]() -> const model::Qualifier & {
+      const model::Qualifier &Qualifier = Qualifiers.back();
+      Qualifiers = Qualifiers.slice(0, Qualifiers.size() - 1);
+      return Qualifier;
+    };
+
+    const auto takeConst = [&]() -> bool {
+      if (not Qualifiers.empty()
+          and Qualifiers.back().Kind() == model::QualifierKind::Const) {
+        Qualifiers = Qualifiers.slice(0, Qualifiers.size() - 1);
+        return true;
+      }
+      return false;
+    };
+
+    // If the set of qualifiers contains any pointers,
+    // the base type does not need to be complete.
+    for (const model::Qualifier &Qualifier : Qualifiers) {
+      if (Qualifier.Kind() == model::QualifierKind::Pointer) {
+        RequireComplete = false;
+        break;
+      }
+    }
+
+    clift::ValueType QualifiedType = rc_recur
```

Comment:
Using `QualifiedType` as variable name is confusing.
`UnwrappedType` maybe?

Reply:
I renamed it to `ResultType`.